internal/servers/destinations.go

```diff
@@ -56,69 +56,67 @@ func (s *DestinationServer) Migrate(ctx context.Context, req *pb.Migrate_Request
 	return &pb.Migrate_Response{}, s.Plugin.Migrate(ctx, tables)
 }
 
+func (*DestinationServer) Write(pb.Destination_WriteServer) error {
+	return status.Errorf(codes.Unimplemented, "method Write is deprecated please upgrade client")
+}
+
 // Note the order of operations in this method is important!
 // Trying to insert into the `resources` channel before starting the reader goroutine will cause a deadlock.
-func (s *DestinationServer) Write(msg pb.Destination_WriteServer) error {
-	resources := make(chan *schema.Resource)
-	var writeSummary *plugins.WriteSummary
+func (s *DestinationServer) Write2(msg pb.Destination_Write2Server) error {
+	resources := make(chan *schema.DestinationResource)
 
 	r, err := msg.Recv()
 	if err != nil {
 		if err == io.EOF {
-			return msg.SendAndClose(&pb.Write_Response{
-				FailedWrites: 0,
-			})
+			return msg.SendAndClose(&pb.Write2_Response{})
 		}
 		return fmt.Errorf("write: failed to receive msg: %w", err)
 	}
-
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		writeSummary = s.Plugin.Write(msg.Context(), r.Source, r.Timestamp.AsTime(), resources)
-	}()
-
-	var resource *schema.Resource
-	if err := json.Unmarshal(r.Resource, &resource); err != nil {
-		close(resources)
-		wg.Wait()
-		return status.Errorf(codes.InvalidArgument, "failed to unmarshal resource: %v", err)
-	}
-	select {
-	case resources <- resource:
-	case <-msg.Context().Done():
-		close(resources)
-		wg.Wait()
-		return msg.Context().Err()
+	var tables schema.Tables
+	if err := json.Unmarshal(r.Tables, &tables); err != nil {
+		return fmt.Errorf("write: failed to unmarshal tables: %w", err)
 	}
+	sourceName := r.Source
+	syncTime := r.Timestamp.AsTime()
 
+	eg := errgroup.Group{}
+	eg.Go(func() error {
+		return s.Plugin.Write(context.Background(), tables, sourceName, syncTime, resources)
+	})
+
+	ctx := msg.Context()
 	for {
 		r, err := msg.Recv()
 		if err != nil {
 			if err == io.EOF {
 				close(resources)
-				wg.Wait()
-				return msg.SendAndClose(&pb.Write_Response{
-					FailedWrites: writeSummary.FailedWrites,
-				})
+				if err := eg.Wait(); err != nil {
+					return fmt.Errorf("failed to wait 1: %w", err)
+				}
+				return msg.SendAndClose(&pb.Write2_Response{})
 			}
 			close(resources)
-			wg.Wait()
-			return fmt.Errorf("write: failed to receive msg: %w", err)
+			if err := eg.Wait(); err != nil {
+				s.Logger.Error().Err(err).Msg("failed to wait 2")
+			}
+			return fmt.Errorf("failed to receive msg: %w", err)
 		}
-		var resource *schema.Resource
+		var resource *schema.DestinationResource
 		if err := json.Unmarshal(r.Resource, &resource); err != nil {
 			close(resources)
-			wg.Wait()
+			if err := eg.Wait(); err != nil {
+				s.Logger.Error().Err(err).Msg("failed to wait 2")
```

Comment:
```suggestion
				s.Logger.Error().Err(err).Msg("failed to wait 3")
```

Looks like the third instance of an error occurring while waiting in this function, we should probably differentiate it in the same way

Reply:
nit: I don't have good wording, but I think rather than numbers it would be better if we can make each error unique through its description, like `failed to wait after reaching io.EOF`

Reply:
Good catch. I think I just forgot this from some of my debugging playground